be/src/olap/memory/partial_row_batch.cpp

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/memory/partial_row_batch.h"
+
+#include "util/bitmap.h"
+
+namespace doris {
+namespace memory {
+
+// Methods for PartialRowBatch
+
+PartialRowBatch::PartialRowBatch(scoped_refptr<Schema>* schema, size_t byte_capacity,
+                                 size_t row_capacity)
+        : _schema(*schema) {
+    _data = reinterpret_cast<uint8_t*>(aligned_malloc(byte_capacity, 64));
+    _bsize = 0;
+    if (!_data) {
+        _byte_capacity = 0;
+        _row_capacity = 0;
+    }
+    _byte_capacity = byte_capacity;
+    _row_capacity = row_capacity;
+}
+
+PartialRowBatch::~PartialRowBatch() {
+    if (_data) {
+        free(_data);
+        _data = nullptr;
+    }
+    _bsize = 0;
+    _byte_capacity = 0;
+    _row_capacity = 0;
+}
+
+const uint8_t* PartialRowBatch::get_row(size_t idx) const {
+    if (!_data || idx >= _row_offsets.size()) {
+        return nullptr;
+    }
+    return _data + _row_offsets[idx] + 4;
+}
+
+// Methods for PartialRowWriter
+
+PartialRowWriter::PartialRowWriter(const Schema& schema)
+        : _schema(schema), _bit_set_size(_schema.cid_size()), _bit_null_size(0) {
+    _temp_cells.resize(_schema.cid_size());
+}
+
+PartialRowWriter::~PartialRowWriter() {}
+
+void PartialRowWriter::start_row() {
+    _bit_null_size = 0;
+    memset(&(_temp_cells[0]), 0, sizeof(CellInfo) * _temp_cells.size());
+}
+
+Status PartialRowWriter::write_row_to_batch(PartialRowBatch* batch) {
+    if (batch->_row_offsets.size() >= batch->row_capacity()) {
+        return Status::InvalidArgument("over capacity");
+    }
+    size_t row_byte_size = byte_size();
+    if (batch->byte_size() + row_byte_size + 4 > batch->byte_capacity()) {
+        return Status::InvalidArgument("over capacity");
+    }
+    *reinterpret_cast<uint32_t*>(batch->_data + batch->_bsize) = row_byte_size;
+    uint8_t* pos = batch->_data + batch->_bsize + 4;
+    RETURN_IF_ERROR(write(&pos));
+    batch->_row_offsets.push_back(batch->_bsize);
+    batch->_bsize = pos - batch->_data;
```

Comment:
Why not construct all items that the batch needs, and construct the batch. If then, the batch can be immutable and this class will not operate batch's private member.

Reply:
Fixed, writer serialize to a buffer, then PartialRowBatch can load/deserialize from this buffer, now PartialRowBatch includes reader functionality